File: server/src/parse-ecb.js
/**
 * Parse ECB eurofxref-daily.xml into a normalized snapshot.
 * Rates are quoted as EUR → currency (units of foreign currency per 1 EUR).
 * EUR itself is always 1.
 */

/**
 * @param {string} xml
 * @returns {{ asOf: string, rates: Record<string, number> }}
 */
export function parseEcbDailyXml(xml) {
  const timeMatch = xml.match(/time=["'](\d{4}-\d{2}-\d{2})["']/);
  if (!timeMatch) {
    throw new Error("ECB XML: missing Cube time attribute");
  }

  const rates = { EUR: 1 };
  const cubeRe = /currency=["']([A-Z]{3})["']\s+rate=["']([0-9.]+)["']/g;
  let m;
  while ((m = cubeRe.exec(xml)) !== null) {
    const code = m[1];
    const rate = Number(m[2]);
    if (!Number.isFinite(rate) || rate <= 0) {
      throw new Error(`ECB XML: invalid rate for ${code}`);
    }
    rates[code] = rate;
  }

  if (Object.keys(rates).length < 2) {
    throw new Error("ECB XML: no currency rates found");
  }

  return { asOf: timeMatch[1], rates };
}

File: app/lib/data/rates/parse_ecb.dart
/// Parse ECB eurofxref-daily.xml into as-of date and rates.
/// Rates are EUR → currency (units of foreign currency per 1 EUR). EUR is always 1.
({String asOf, Map<String, double> rates}) parseEcbDailyXml(String xml) {
  final timeMatch = RegExp(r'''time=["'](\d{4}-\d{2}-\d{2})["']''').firstMatch(xml);
  if (timeMatch == null) {
    throw FormatException('ECB XML: missing Cube time attribute');
  }

  final rates = <String, double>{'EUR': 1};
  final cubeRe = RegExp(
    r'''currency=["']([A-Z]{3})["']\s+rate=["']([0-9.]+)["']''',
  );
  for (final m in cubeRe.allMatches(xml)) {
    final code = m.group(1)!;
    final rate = double.tryParse(m.group(2)!);
    if (rate == null || !rate.isFinite || rate <= 0) {
      throw FormatException('ECB XML: invalid rate for $code');
    }
    rates[code] = rate;
  }

  if (rates.length < 2) {
    throw FormatException('ECB XML: no currency rates found');
  }

  return (asOf: timeMatch.group(1)!, rates: rates);
}
